### PR TITLE
Pro release v33

### PR DESCRIPTION
**Pull request by renanrodrigo**
This changeset represents the diff to be uploaded as version `33` of the `ubuntu-pro-client` package.

* **Pro Client Representative:** @renanrodrigo 
* **Sponsor Reviewer:** @lucaskanashiro 
* **SRU Reviewer:** @basak

### CODE REVIEW

**Comment by renanrodrigo** [comment]
Our autopkgtests are superficial, but overall they are all passing for:
- Bionic: [amd64](https://autopkgtest.ubuntu.com/results/autopkgtest-bionic-ua-client-staging/bionic/amd64/u/ubuntu-advantage-tools/20240613_201032_62e85@/log.gz) [arm64](https://autopkgtest.ubuntu.com/results/autopkgtest-bionic-ua-client-staging/bionic/arm64/u/ubuntu-advantage-tools/20240613_201529_a84fd@/log.gz) [armhf](https://autopkgtest.ubuntu.com/results/autopkgtest-bionic-ua-client-staging/bionic/armhf/u/ubuntu-advantage-tools/20240613_195839_e4468@/log.gz) [i386](https://autopkgtest.ubuntu.com/results/autopkgtest-bionic-ua-client-staging/bionic/i386/u/ubuntu-advantage-tools/20240613_202009_9ca6a@/log.gz) [ppc64el](https://autopkgtest.ubuntu.com/results/autopkgtest-bionic-ua-client-staging/bionic/ppc64el/u/ubuntu-advantage-tools/20240613_201049_d7ba7@/log.gz) [s390x](https://autopkgtest.ubuntu.com/results/autopkgtest-bionic-ua-client-staging/bionic/s390x/u/ubuntu-advantage-tools/20240613_202800_b2799@/log.gz)
- Focal: [amd64](https://autopkgtest.ubuntu.com/results/autopkgtest-focal-ua-client-staging/focal/amd64/u/ubuntu-advantage-tools/20240613_201657_f614e@/log.gz) [arm64](https://autopkgtest.ubuntu.com/results/autopkgtest-focal-ua-client-staging/focal/arm64/u/ubuntu-advantage-tools/20240613_200953_54e05@/log.gz) [armhf](https://autopkgtest.ubuntu.com/results/autopkgtest-focal-ua-client-staging/focal/armhf/u/ubuntu-advantage-tools/20240613_195707_fe571@/log.gz) [ppc64el](https://autopkgtest.ubuntu.com/results/autopkgtest-focal-ua-client-staging/focal/ppc64el/u/ubuntu-advantage-tools/20240613_200203_753b1@/log.gz) [s390x](https://autopkgtest.ubuntu.com/results/autopkgtest-focal-ua-client-staging/focal/s390x/u/ubuntu-advantage-tools/20240613_200407_2e1f8@/log.gz)
- Jammy: [amd64](https://autopkgtest.ubuntu.com/results/autopkgtest-jammy-ua-client-staging/jammy/amd64/u/ubuntu-advantage-tools/20240613_195440_a9fff@/log.gz) [arm64](https://autopkgtest.ubuntu.com/results/autopkgtest-jammy-ua-client-staging/jammy/arm64/u/ubuntu-advantage-tools/20240613_200754_706b9@/log.gz) [armhf](https://autopkgtest.ubuntu.com/results/autopkgtest-jammy-ua-client-staging/jammy/armhf/u/ubuntu-advantage-tools/20240613_200029_2f58b@/log.gz) [ppc64el](https://autopkgtest.ubuntu.com/results/autopkgtest-jammy-ua-client-staging/jammy/ppc64el/u/ubuntu-advantage-tools/20240613_200036_a60e6@/log.gz) [s390x](https://autopkgtest.ubuntu.com/results/autopkgtest-jammy-ua-client-staging/jammy/s390x/u/ubuntu-advantage-tools/20240613_202358_4e158@/log.gz)
- Noble: [amd64](https://autopkgtest.ubuntu.com/results/autopkgtest-noble-ua-client-staging/noble/amd64/u/ubuntu-advantage-tools/20240613_200028_017d6@/log.gz) [arm64](https://autopkgtest.ubuntu.com/results/autopkgtest-noble-ua-client-staging/noble/arm64/u/ubuntu-advantage-tools/20240613_201313_a91e3@/log.gz) [armhf](https://autopkgtest.ubuntu.com/results/autopkgtest-noble-ua-client-staging/noble/armhf/u/ubuntu-advantage-tools/20240614_012019_e91ff@/log.gz) [ppc64el](https://autopkgtest.ubuntu.com/results/autopkgtest-noble-ua-client-staging/noble/ppc64el/u/ubuntu-advantage-tools/20240613_200346_9e807@/log.gz) [s390x](https://autopkgtest.ubuntu.com/results/autopkgtest-noble-ua-client-staging/noble/s390x/u/ubuntu-advantage-tools/20240613_195814_baeb0@/log.gz)
- Oracular: [amd64](https://autopkgtest.ubuntu.com/results/autopkgtest-oracular-ua-client-staging/oracular/amd64/u/ubuntu-advantage-tools/20240613_204019_1e6a2@/log.gz) [arm64](https://autopkgtest.ubuntu.com/results/autopkgtest-oracular-ua-client-staging/oracular/arm64/u/ubuntu-advantage-tools/20240613_201524_4edd9@/log.gz) [armhf](https://autopkgtest.ubuntu.com/results/autopkgtest-oracular-ua-client-staging/oracular/armhf/u/ubuntu-advantage-tools/20240613_195358_b0c55@/log.gz) [ppc64el](https://autopkgtest.ubuntu.com/results/autopkgtest-oracular-ua-client-staging/oracular/ppc64el/u/ubuntu-advantage-tools/20240613_200615_af5b9@/log.gz) [s390x](https://autopkgtest.ubuntu.com/results/autopkgtest-oracular-ua-client-staging/oracular/s390x/u/ubuntu-advantage-tools/20240613_195116_6bef0@/log.gz)

For Xenial, we have passing autopkgtests for [amd64](https://autopkgtest.ubuntu.com/results/autopkgtest-xenial-ua-client-staging/xenial/amd64/u/ubuntu-advantage-tools/20240613_201038_b9b17@/log.gz) [armhf](https://autopkgtest.ubuntu.com/results/autopkgtest-xenial-ua-client-staging/xenial/armhf/u/ubuntu-advantage-tools/20240613_201631_cdab1@/log.gz) [ppc64el](https://autopkgtest.ubuntu.com/results/autopkgtest-xenial-ua-client-staging/xenial/ppc64el/u/ubuntu-advantage-tools/20240613_202724_c3144@/log.gz) and [s390x](https://autopkgtest.ubuntu.com/results/autopkgtest-xenial-ua-client-staging/xenial/s390x/u/ubuntu-advantage-tools/20240613_201458_12ff8@/log.gz), but are failing to even start tests on [arm64](https://autopkgtest.ubuntu.com/results/autopkgtest-xenial-ua-client-staging/xenial/arm64/u/ubuntu-advantage-tools/20240614_012454_7d0a0@/log.gz) and [i386](https://autopkgtest.ubuntu.com/results/autopkgtest-xenial-ua-client-staging/xenial/i386/u/ubuntu-advantage-tools/20240614_012756_24dc6@/log.gz) - seems there is some problem in the infra itself. Let me know if you think that is a problem we need to solve our side.

**Comment by renanrodrigo** [comment]
I didn't care about Mantic as we are not releasing there. The estimated date for this release is end of July, when Mantic will be EOL

**Comment by renanrodrigo** [comment]
@lucaskanashiro FYI, this branch received (and may still receive) a couple test fixes and adjustments - no new functionality will be introduced though, we are just making sure all is good.

File: debian/po/pt_BR.po [hunk 1]
```diff
@@ -351,6 +351,13 @@ msgid "Finished upgrade of Ubuntu Pro service configuration"
 msgstr "Atualização das configurações de serviço do Ubuntu Pro concluída"
 
 #: ../../uaclient/messages/__init__.py:85
+#, python-brace-format
+msgid ""
+"Detaching Ubuntu Pro. Previously attached subscription was only valid for "
+"Ubuntu {release} ({series_codename}) release."
+msgstr ""
```
Comment (lucaskanashiro) [inline debian/po/pt_BR.po:358]: Some new strings are added to PO files, do we want to also ship pt_BR translations for that? Not sure about the Pro team process, but I'd expect to have all new additions translated during the release process.

**Review by lucaskanashiro — APPROVED** [review]
I added a couple of pretty simple comments, which are definitely not blocking the release IMHO. Please, take a look at them and see if they make sense, if you prefer they could get fixed in the next release. I am already approving this PR.

File: debian/po/pt_BR.po (hunk 1 above)
Comment (orndorffgrant) [reply debian/po/pt_BR.po:358]: Our policy is to be pretty relaxed about adding new strings and deleting existing strings.
We should put in a best effort for updating translations for strings that were already translated but got changed, but not block on it.

File: features/ubuntu_pro.feature [hunk 2]
```diff
@@ -224,7 +224,7 @@ Feature: Command behaviour when auto-attached in an ubuntu PRO image
       """
       "attached": false
       """
-    When I run `cat /var/log/ubuntu-advantage.log` with sudo
+    When I run `journalctl` as non-root
```
[the remark this answers is not in the document]
Comment (orndorffgrant) [reply features/ubuntu_pro.feature:227]: Yes that would be better. I'll go ahead and make that change

[the remark this answers is not in the document]
Comment (orndorffgrant) [reply features/ubuntu_pro.feature]: Done!

**Comment by basak** [comment]
I reviewed up to upstream commit cc151998, starting from upstream commit be597708 (tagged 32.3 upstream). Oracular is on 32.3.1. pkg/import/32.3 from git-ubuntu is identical to upstream tag 32.3, and git-ubuntu has further rich history with 2b115841 and e93055b0 being the material changes for 32.3.1. These I see in be597708..cc151998 as 95d11867 and 58d49cad respectively. git range-diff reports these commits as identical, so therefore apart from these two commits upstream, everything in be597708..cc151998 are the changes that would affect Ubuntu.

This was a bit of a pain to identify, FWIW. I wonder if we could make the process better next time? I don't mind where it is but what I want to review is precisely the set of (rich) git commits that will change from an Ubuntu user's perspective.

[Not for this release] 31dbcc73 fips: dynamically unhold packages to upgrade during auto-upgrade

I'm curious to understand why apt holds are being used for FIPS rather than apt pinning. This is surprising to me, but probably not a topic for this PR.

[Needs Information] 5bb896cb apt: use version string for comparison instead of version object

$ dpkg --compare-versions 1.2-ubuntu3.4 eq 1.2-ubuntu3.04 && echo equal
equal

So version string comparisons cannot be used to determine equality in the general case. If such a version string gets normalised in some places but not others, then we would end up with a mismatch.

What's the reason for this change? Am I missing something? This is an example of why rationale should be in commit messages ☺

[Needs Information] fe6b23d4 Updated attach to check for an onlySeries field in the contract response

Hardcoding CODENAME_TO_RELEASE has some long term implications. What happens when the development release is released? Will we then be required to SRU an update to add that version? Or, if prepared before release, it's yet another thing to change if we ever adjust our release schedule. Imagine if the release were to get delayed by a month and then everything would be further blocked on an ubuntu-advantage-tools release fixing this data! Instead, the ubuntu-pro-client binary package already depends on distro-info, and this information is available to you there. Is there a good reason this cannot be used instead?

[Comment only] 255d5a23 Add function to create timed job for checking release every 24 hours and check release on release-upgrade

It seems odd to me to do this on a timed job rather than on reboot. However, the latter would mean yet another service, so perhaps that's the reason? Since the timed job is only active when attached though, I don't think this is much of a concern from an SRU perspective.

**Comment by lucaskanashiro** [comment]
Well, based on Robie's comment, I believe the Pro client review process for PRs should also include checking for descriptive git commit messages. Good commit messages (describing how/what/why the changes are needed) would smooth the release process for everyone (reviewers will not seek for clarifications, and developers will not need to spend time explaining reviewers what they have in their minds). In this PR specifically, Robie could have approved (or requested changes) straightway if the information was there, reducing the back and forth that usually happens. Just my 2 cents.

**Comment by renanrodrigo** [comment]
@basak 

> This was a bit of a pain to identify, FWIW. I wonder if we could make the process better next time? I don't mind where it is but what I want to review is precisely the set of (rich) git commits that will change from an Ubuntu user's perspective.

We do have a better process, but this didn't follow the process as it's quite new. This was a separate upload done by @panlinux to hotfix an apparmor issue, and the new version will supersede the current one (we don't even mention it in the changelog). We agreed on that with Andreas so we could follow on with the release, and he'd have the freedom to upload the bugfix in the meantime. We could have anticipated and made it explicit in the PR though - a communication issue, we're sorry.

> https://github.com/canonical/ubuntu-pro-client/commit/31dbcc73181d1e52e1de7a3e44346e0157279b75 fips: dynamically unhold packages to upgrade during auto-upgrade
I'm curious to understand why apt holds are being used for FIPS rather than apt pinning. This is surprising to me, but probably not a topic for this PR.

It's easy - you can see we use the PIN for FIPS, and we never hold packages - we only unhold them (and then they are subject to the pinning we do). The reason FIPS packages may be pre-held is that CPC is creating FIPS cloud images without making use of the Pro Client in the process, and this is their solution. Further discussion welcome on a separate channel if you are interested in more.

>  I realised during the review that the issue here was fixed in a subsequent commit, but forgot to remove the comment, so I'm doing that now. Review time required is a good reason to follow a rebase workflow rather than a fixup workflow though for PR review iterations!]

We do have a rebase workflow, or rather a squash workflow, when issues and fixes are present in the same PR. This one was either a whoops we forgot to do it, OR it slipped through and we found it in a subsequent PR. Again, sorry for the inconvenience.

**Comment by orndorffgrant** [comment]
> [Needs Information] [5bb896c](https://github.com/canonical/ubuntu-pro-client/commit/5bb896cbd8b2ee3ae070c2f8f09c6c8903f19844) apt: use version string for comparison instead of version object
> 
> $ dpkg --compare-versions 1.2-ubuntu3.4 eq 1.2-ubuntu3.04 && echo equal equal
> 
> So version string comparisons cannot be used to determine equality in the general case. If such a version string gets normalised in some places but not others, then we would end up with a mismatch.
> 
> What's the reason for this change? Am I missing something? This is an example of why rationale should be in commit messages ☺

It was a review suggestion as a sort of easy micro-optimization. We didn't realize equality couldn't be checked via string comparison (but of course now seeing the example, it makes sense that it can't). I'll revert it. Thanks for catching it!

> [Comment only] [255d5a2](https://github.com/canonical/ubuntu-pro-client/commit/255d5a23d5b0f2042c16c2c5fe1714ea08c58495) Add function to create timed job for checking release every 24 hours and check release on release-upgrade
> 
> It seems odd to me to do this on a timed job rather than on reboot. However, the latter would mean yet another service, so perhaps that's the reason? Since the timed job is only active when attached though, I don't think this is much of a concern from an SRU perspective.

A timer was the initial suggestion in the [spec](https://docs.google.com/document/d/1lMos-aBpg59xDQaduF1z2FilluXLdaId5skXNPIoFOQ/edit] and was accepted because of the urgency of the feature and the ease of implementation. You're right though that it doesn't need to run except for on reboot. I've opened an issue to optimize this: #3189

**Comment by basak** [comment]
Reviewed from my previous review point (cc151998) up to current `main` (1d72b42f) and all looks good. Thanks!

**Review by basak — APPROVED** [review]

**Comment by renanrodrigo** [comment]
Thanks much @basak!
@lucaskanashiro shall we start the upload party?

**Comment by renanrodrigo** [comment]
@lucaskanashiro if you could please upload the contents of the `main` branch to Oracular. Other branches have been created with backport entries for the `unapproved` uploads too: [Xenial](https://github.com/canonical/ubuntu-pro-client/tree/release-33-xenial), [Bionic](https://github.com/canonical/ubuntu-pro-client/tree/release-33-bionic), [Focal](https://github.com/canonical/ubuntu-pro-client/tree/release-33-focal), [Jammy](https://github.com/canonical/ubuntu-pro-client/tree/release-33-jammy), [Noble](https://github.com/canonical/ubuntu-pro-client/tree/release-33-noble).

**Comment by lucaskanashiro** [comment]
The content of `main` was just uploaded to Oracular. If everything goes well as expected, I'll upload the SRUs tomorrow.

**Comment by lucaskanashiro** [comment]
The SRU uploads were also done.

**Comment by renanrodrigo** [comment]
This branch now contains the Pro Client version `33.1`. The diff from `33` contains:
- a commit fixing a test
- the contents of https://github.com/canonical/ubuntu-pro-client/pull/3196

@lucaskanashiro if you could please verify the changes are good and upload this version to oracular, and to the unapproved queues;
@basak we will ping you once the upload happens for acceptance, so we can start the verification of this version,

**Comment by lucaskanashiro** [comment]
@renanrodrigo upload to Oracular is done. The SRU branches are not up-to-date, let me know once they are and I can upload the SRUs.

**Comment by renanrodrigo** [comment]
@lucaskanashiro all branches now ready. Please remember to keep both changelog entries in the upload (i.e. 33 and 33.1)

**Comment by lucaskanashiro** [comment]
SRUs uploaded.

File: uaclient/version.py [hunk 3]
```diff
@@ -15,7 +15,7 @@
 from uaclient.exceptions import ProcessExecutionError
 from uaclient.system import subp
 
-__VERSION__ = "32.3"
```
Comment (renanrodrigo) [inline uaclient/version.py:18]: @lucasmoura keep in mind that youmerged the changelog PR but are missing the version bump to 33.2

**Comment by orndorffgrant** [comment]
While not officially released to -updates. 33.2 is done and verified. It is only held up by a bug in a dependency that must be fixed. As such, we are calling this release done.

**Pull request closed without merging**